src/background.c

```diff
@@ -0,0 +1,292 @@
+#define _GNU_SOURCE 1
+
+#include "config.h"
+
+#include <errno.h>
+#include <locale.h>
+#include <string.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+
+#include <gtk/gtk.h>
+
+#include <glib/gi18n.h>
+
+#include "xdg-desktop-portal-dbus.h"
+#include "shell-dbus.h"
+
+#include "background.h"
+#include "fdonotification.h"
+#include "request.h"
+#include "utils.h"
+
+static OrgGnomeShellIntrospect *shell;
+
+typedef enum { BACKGROUND, RUNNING, ACTIVE } AppState;
+
+static gboolean
+handle_get_app_state (XdpImplBackground *object,
+                      GDBusMethodInvocation *invocation)
+{
+  g_autoptr(GVariant) windows = NULL;
+  g_autoptr(GHashTable) app_states = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+  g_autoptr(GError) error = NULL;
+  GVariantBuilder builder;
+  GHashTableIter iter;
+  const char *key;
+  gpointer value;
+
+  g_debug ("background: handle GetAppState");
+
+  if (!org_gnome_shell_introspect_call_get_windows_sync (shell, &windows, NULL, &error))
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             XDG_DESKTOP_PORTAL_ERROR,
+                                             XDG_DESKTOP_PORTAL_ERROR_FAILED,
+                                             "Could not get window list: %s", error->message);
+      return TRUE;
+    }
+
+  if (windows)
+    {
+      g_autoptr(GVariantIter) iter = g_variant_iter_new (windows);
+      GVariant *dict;
+
+      while (g_variant_iter_loop (iter, "{t@a{sv}}", NULL, &dict))
+        {
+          const char *app_id = NULL;
+          char *app;
+          gboolean hidden = FALSE;
+          gboolean focus = FALSE;
+          AppState state = BACKGROUND;
+
+          g_variant_lookup (dict, "app-id", "&s", &app_id);
+          g_variant_lookup (dict, "is-hidden", "b", &hidden);
+          g_variant_lookup (dict, "has-focus", "b", &focus);
+
+          if (app_id == NULL)
+            continue;
+
+          app = g_strdup (app_id);
+          if (g_str_has_suffix (app, ".desktop"))
+            app[strlen (app) - strlen (".desktop")] = '\0';
```

Comment:
This here is problematic. Flatpaks can validly have an app id that ends with .desktop. And in fact, `org.telegram.desktop` does this, which has triggered bugs in a bunch of places like this..

Reply:
I think it's a hack until the gnome-shell bug mentioned in https://github.com/flatpak/xdg-desktop-portal-gtk/pull/198#issuecomment-492026996 is fixed.

Reply:
Yeah, sorry, should have commented on that. It is indeed a temporary hack